vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go

```diff
@@ -0,0 +1,208 @@
+			continue
+		}
+
+		updatedContainerResources, err := getRecommendationForContainerWithRestrictedScaling(*container, &containerRecommendation, replicas, lowWatermark, highWatermark)
+		if err != nil {
+			return nil, fmt.Errorf("cannot update recommendation for container name %v", container.Name)
+		}
+		updatedRecommendations = append(updatedRecommendations, *updatedContainerResources)
+	}
+
+	return &vpa_types.RecommendedPodResources{ContainerRecommendations: updatedRecommendations}, nil
+}
+
+// getRecommendationForContainerWithRestrictedScaling returns a recommendation for the given container, adjusted to obey maintainedRatios policy
+func getRecommendationForContainerWithRestrictedScaling(
+	container apiv1.Container,
+	containerRecommendation *vpa_types.RecommendedContainerResources,
+	replicas int,
+	lowWatermark, highWatermark int) (*vpa_types.RecommendedContainerResources, error) {
+
+	amendedRecommendations := containerRecommendation.DeepCopy()
+
+	isRestricted := isRestrictedScaling(containerRecommendation, container, lowWatermark, highWatermark, replicas)
+	process := func(recommendation apiv1.ResourceList) {
+		applyRestrictedScalingVPAPolicy(recommendation, isRestricted, container.Resources.Requests)
+	}
+
+	process(amendedRecommendations.Target)
+	// Do we want to change LowerBound and UpperBound?
+
+	return amendedRecommendations, nil
+}
+
+// isRestrictedScaling: detects if the recommendation is a restricted scaling
+func isRestrictedScaling(containerRecommendation *vpa_types.RecommendedContainerResources, container apiv1.Container, lowWatermark, highWatermark int, replicas int) bool {
+	isAnUpscale := isUpscale(containerRecommendation.Target, container.Resources.Requests)
+	isADownscale := isDownscale(containerRecommendation.Target, container.Resources.Requests)
+	isRestricted := false
+	if isAnUpscale && replicas <= highWatermark {
+		isRestricted = true
+	}
+	if isADownscale && replicas >= lowWatermark {
+		isRestricted = true
+	}
+	if isRestricted {
+		klog.V(2).Infof("[%s] Scaling is restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	} else {
+		klog.V(4).Infof("[%s] Scaling is not restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	}
+	return isRestricted
+}
+
+// isUpscale: detects if the recommendation is an upscale. Any increase of resources is considered as upscale.
+func isUpscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
+	for recommendationType, _ := range recommendation {
+		q := containerOriginalResources.Name(recommendationType, resource.DecimalSI).MilliValue()
+		r := recommendation.Name(recommendationType, resource.DecimalSI).MilliValue()
+		if r > q {
+			return true
+		}
```

Comment:
Nit: you can use `return r.Cmp(q) == 1` instead.

Reply:
I never understand how `Cmp` works, I find a basic comparison easier to read